rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,240 @@
+# RESTful Service Specifications
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## Goals
+
+ - Uniformity between data-centric APIs and remote procedure APIs
+ - Use of pre-existing standards/conventions when possible
+ - Minimization of bike-shedding
+
+Standards/Conventions used as inspiration:
+ - [Whitehouse API Standards](https://github.com/WhiteHouse/api-standards)
+  - RESTful URLs
+ - [Spring Data REST](http://docs.spring.io/spring-data/rest/docs/current/reference/html/)
+  - paging conventions
+  - uses [JSON HAL](https://tools.ietf.org/html/draft-kelly-json-hal-08)
+ - [JSend](http://labs.omniti.com/labs/jsend)
+  - API response conventions
+
+## URIs
+
+Version granularity is up to implementation. However, the base URI for an
+entity must include a version.
+
+The base URI for a given resource type is based on the *singular* name of the
+resource. Example:
+
+    https://example.com/rest/v1/user
+
+## Operations Against a Resource Type
+
+The following operations may be supported:
+
+| HTTP Operation | Description            | Query Parameters              | Request Body | Response Body  |
+|----------------|------------------------|-------------------------------|--------------|----------------|
+| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
+| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
+| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
+| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
+| `DELETE`       | Delete ALL resources   | -                             | -            | -              |
+
+Endpoints may not support all methods (i.e. a service may provide read-only
+access to a resource, or a service may not support certain operations for
+performance or other considerations; for example, `GET` may not be supported
+for `user` if listing all users is disallowed for performance reasons.
+
+## Operations Against Specific Resources
+Specific resources will be addressed through the base URI plus the identifier
+for that resource.
+
+    /user/{id}
```

Comment:
First off, loving the discussion!

> Query params are for filtering. Path params are for navigation and reference.

Unfortunately, I don't think we can rely on these kinds of absolutes. We have to make compromises. REST implementations widely vary in practice. This is unfortunate, but c'est la vie.

> IMO, APIs are not for business users, they are for developers and the services/applications they write.

I think this line of reasoning can be risky long term. It is good to strive for [ubiquitous language](http://martinfowler.com/bliki/UbiquitousLanguage.html) between developers, QA, and business users. If we speak the same language, there is less lost in translation. The identifiers for a domain "thing" can be important, so I think it's not unreasonable to ask everyone to speak the same language here, though maybe not critical for ids. I would empathize with QA here particularly.

> GET path?x=foo&y=bar&z=car. For every resource that is navigable via query params only, the consumer has to determine for each combination of fields: Am I specifying search criteria, identifying a single resource, or multiple resources?

I think this is a good point, but I think maybe solving is not so bad. How about this: the answer is you are always identifying a single resource just like with a path param. We have in the spec that requests against /resource are for everything in that resource. We have in the spec that /resource/{function} can do anything. So that means anything else is one entity and one entity only, just like /user/{id} is. If not enough params are provided for a compound identity then a 4xx error response is returned according to error schema (below). GET,PUT,PATCH should work as expected: one entity or fail. There is no ambiguity and the spec is consistent.

As far as having to know what query params to use, I agree it is definitely easier when everything is just a single path param id. Taking inspiration from @bserdar 's suggestion, maybe /user/{id} could be sugar for /user?id={id}. Then, if "id" is a param, it just works. If it isn't, you get a descriptive error that says what params need to be provided. What do you think?

Additionally, as far as consumer discoverability goes, I hope that whatever solution we land on has a clear route for good, auto-generated clients in many languages. This makes some of the details of the REST interface not as important to get perfect since most would see it through a language-specific client anyway.

My 2c. Good discussion!